### DISCUSSION TITLE
bug: Preprocessor definitions

### DISCUSSION THREAD

**Discussion by jphaupt** [krvajal/vscode-fortran-support#714] [Q&A - deprecated see Fortran Discourse]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Sorry if I missed this somewhere. For your convenience, I made a little repo for this issue (+ another issue): https://github.com/jphaupt/min_cmake_preproc/

Let's say I have a preprocessor definition `NUMBER_PREPROCESSED=real`. I then want to use in my code `NUMBER_PREPROCESSED(dp)`. It looks like this is actually picked up correctly, as when you hover over the line it shows you `#define NUMBER_PREPROCESSED real`, but it still shows an error, and more importantly it deletes the .mod file (if there was one and if this is a module), which ruins indexing if the module is embedded in a bigger project. This is however valid code, as you can do `target_compile_definitions(<target> PUBLIC NUMBER_PREPROCESSED=real)` in cmake. 

### Expected Behaviour

I expect it to notice the definition, as it does correctly, but to not show an error and not to delete the .mod file. 

### Version of Modern Fortran

v3.2.0

### Version of Visual Studio Code

1.72.2

### Platform and Architecture

Linux

**Comment by gnikit** [comment]
There are 2 different things that are going on. 
1. The language server (fortls) will probably, and from what you are saying it does so successfully, parse your documents for preprocessor macros.
2. The linter will run your code through the actual compiler by default gfortran and give you compiler-time diagnostics.

What you are probably seeing is your linter not knowing about your project configuration and your preprocessor macros. There is no current integration with CMake see https://github.com/fortran-lang/vscode-fortran-support/issues/133

**Reply by gnikit** [reply]
You will have to set the settings to the linter manually

**Reply by jphaupt** [reply]
Thanks for the fast response. I saw that issue but thought that was different. I don't need it to pick up on my CMake configurations; I just want it to directly take the definition I gave it with the .vscode/settings.json file, i.e. I am setting it manually (not looking for plug-and-play, as you say in the other issue).

**Reply by jphaupt** [reply]
> You will have to set the settings to the linter manually

I'm sorry, isn't that what I am doing with settings.json? Can you explain how to do that?

**Reply by gnikit** [reply]
> I'm sorry, isn't that what I am doing with settings.json? Can you explain how to do that?

No, you are setting the settings for the language server. There is an entire different set of settings for the linter. You can see all the available options the extension offers in (Extensions> Search for Modern Fortran > Feature Contributions)
![image](https://user-images.githubusercontent.com/16143716/196243278-7384fd81-61df-4037-8e9e-154b0d43c0c3.png)

As for this

> it deletes the .mod file (if there was one and if this is a module), which ruins indexing if the module is embedded in a bigger project.

Modern Fortran does not do that. It can override it if you haven't set the linter output module directory to something reasonable but it should not delete files.

**Reply by jphaupt** [reply]
Ah that's embarrassing. Thanks, on the off chance that anyone else ends up here, the setting is `fortran.linter.extraArgs`.
```
    "fortran.linter.extraArgs": [
        "-DNUMBER_PREPROCESSED=real"
    ]
```
No idea what's removing the .mod file, but it happens even if Modern Fortran is the only extension I have enabled (along with the C/C++ extension which it depends on, of course).

**Reply by gnikit** [reply]
> Ah that's embarrassing. Thanks, on the off chance that anyone else ends up here, the setting is `fortran.linter.extraArgs`.

No worries. At some point I we will have to make a documentation page, but that a bit of work so we will probably need some funding for that (probable ETA summer 2023). At some point we will have to figure out how to merge the `fortls` settings with the linter settings, since it doesn't make much sense to have to define them twice.

> No idea what's removing the .mod file, but it happens even if Modern Fortran is the only extension I have enabled (along with the C/C++ extension which it depends on, of course).

No idea either. Can you set the logging verbosity level to Debug and post the logs? The other thing you can try to make sure that Modern Fortran does not mess with your build is create a tmp dir e.g. `.vscode/fortran` and set that as the linter mod output.
